kernel/ps2_controller/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+#![no_std]
+
+use log::{debug, warn};
+use ps2::{write_command, flush_output_buffer, read_config, write_config, read_controller_test_result, read_port_test_result, PortTestResult, HostToControllerCommand::*, reset_keyboard, reset_mouse};
+
+
+// see https://wiki.osdev.org/%228042%22_PS/2_Controller#Initialising_the_PS.2F2_Controller
+/// initialize the first and second (if it exists) PS/2 port
+pub fn init() -> Result<(), &'static str> {
+    // Step 1: Initialise USB Controllers
+    // no USB support yet
+
+    // Step 2: Determine if the PS/2 Controller Exists
+    let acpi_tables = acpi::get_acpi_tables().lock();
+    if let Some(fadt) = fadt::Fadt::get(&acpi_tables) {
+        // if earlier than ACPI v2, PS/2 controller is assumed to exist
+        if fadt.header.revision > 1 {
+            let has_controller = fadt.iapc_boot_architecture_flags & 0b10 == 0b10;
+            if !has_controller {
+                //TODO: sounds more like this should be done before calling this function
+                return Err("no PS/2 Controller present");
+            }
+        }
+    }
+
+    // Step 3: Disable Devices
+    write_command(DisablePort1);
+    write_command(DisablePort2);
+
+    // Step 4: Flush The Output Buffer
+    flush_output_buffer();
+
+    // Step 5: Set the Controller Configuration Byte
+    let mut config = read_config();
+    config.set_port1_interrupt_enabled(false);
+    config.set_port2_interrupt_enabled(false);
+    config.set_port1_translation_enabled(false);
+    let has_mouse = config.port2_clock_disabled();
+    write_config(config.clone());
+
+    // Step 6: Perform Controller Self Test
+    write_command(TestController);
+    read_controller_test_result()?;
+    debug!("passed PS/2 controller test");
+    // as this can reset the controller on some hardware, we restore the config
+    write_config(config.clone());
+
+    // Step 7 not needed, already done above
```

Comment:
can you explicitly state what Step 7 is, even though it's already been done? It'd be nice to preserve it if the osdev wiki goes down